07-errors.md

```diff
@@ -0,0 +1,284 @@
+# Error Conditions
+
+This document specifies the error conditions in the FVM and the associated codes.
+
+## Exit Codes
+
+Exit codes are numeric, 32bit integer codes that are returned when an actor exits both successfully and in error.
+Exit codes appear on-chain inside the message receipt, and therefore form part of consensus.
+
+We classify exit codes as "normal" and "abnormal".
+
+The only _normal_ exit code is `0`, or `Ok`. The `Ok` exit code indicates that the message was
+applied successfully.
+
+All other exit codes are _abnormal_ and indicate that the message _was not_ applied and that the
+chain state _was not_ updated.
+
+### System "Abnormal Exit" Codes
+
+System abnormal exit codes are generated by the VM itself to indicate that the message itself failed
+to apply. These exit codes will never be observed or returned directly by actors.
+
+| Code  | Status                     | Description                                               |
+|-------|----------------------------|-----------------------------------------------------------|
+| 1     | `SysErrSenderInvalid`      | sender doesn't exist                                      |
+| 2     | `SysErrSenderStateInvalid` | sender's message shouldn't have been included             |
+| 3     | reserved                   |                                                           |
+| 4     | `SysErrActorPanic`         | message receiver trapped                                  |
```

Comment:
This one makes me a bit nervous. There's none such in the existing exit codes because a panic may not be replicated by other nodes. Is it ok now because we have a real VM that should be 100% deterministic, even wrt to resource limits etc?

Note this down below in "changes from pre-FVM".

Reply:
This situation can happen in WASM if, for example, the code goes into an [`unreachable`](https://developer.mozilla.org/en-US/docs/WebAssembly/Reference/Control_flow/unreachable) instruction, without explicitly signalling an abort. `Panic` isn't quite the right term here, though, because of its associations with high-level language constructions. I'd imagine something more abstract like `SysErrExecutionTerminated` being more akin.

Reply:
"Terminated" is a bit ambiguous. Maybe `SysErrIllegalInstruction`? `SysErrActorCrash`?

Reply:
`SysErrIllegalInstruction`

+1